include/RAJA/policy/hip/MemUtils_HIP.hpp

```diff
@@ -299,205 +301,240 @@ hipDeviceProp_t& device_prop()
+    data.func_max_blocks_per_sm = hip::device_prop().maxThreadsPerMultiProcessor/1024;
+    if (data.func_max_blocks_per_sm <= 0) { data.func_max_blocks_per_sm = 1 }
 #endif
 
-    if (data.multiProcessorCount == uninitialized) {
-
-      data.multiProcessorCount = hip::device_prop().multiProcessorCount;
-
-    }
-
-    data.max_blocks *= data.multiProcessorCount;
-
-    data.prev_shmem_size = shmem_size;
 
   }
 
-  max_blocks = data.max_blocks;
-
+  return data;
 }
 
-struct HipOccMaxBlocksVariableThreadsData
-{
-  size_t prev_shmem_size;
-  int prev_num_threads;
-  int max_blocks;
-  int multiProcessorCount;
-};
-
-template < typename RAJA_UNUSED_ARG(UniqueMarker), typename Func >
+//! Get the maximum occupancy of a kernel with runtime threads per block
+template < typename RAJA_UNUSED_ARG(UniqueMarker) >
 RAJA_INLINE
-void hip_occupancy_max_blocks(Func&& func, size_t shmem_size,
-                               int &max_blocks, int num_threads)
+HipOccMaxBlocksData hip_occupancy_max_blocks(const void* func,
+    size_t func_dynamic_shmem_per_block, int func_threads_per_block)
 {
-  static constexpr int uninitialized = 0;
-  static constexpr size_t uninitialized_size_t = std::numeric_limits<size_t>::max();
-  static thread_local HipOccMaxBlocksVariableThreadsData data = {
-      uninitialized_size_t, uninitialized, uninitialized, uninitialized};
+  static thread_local HipOccMaxBlocksData data {
+      std::numeric_limits<size_t>::max(),
+      -1,
+      hip::device_prop().multiProcessorCount,
+      hip::device_prop().maxThreadsPerMultiProcessor,
+      -1 };
+
+  if ( data.func_dynamic_shmem_per_block != func_dynamic_shmem_per_block ||
+       data.func_threads_per_block != func_threads_per_block ) {
 
-  if ( data.prev_shmem_size  != shmem_size ||
-       data.prev_num_threads != num_threads ) {
+    data.func_dynamic_shmem_per_block = func_dynamic_shmem_per_block;
+    data.func_threads_per_block = func_threads_per_block;
 
 #ifdef RAJA_ENABLE_HIP_OCCUPANCY_CALCULATOR
     hipErrchk(hipOccupancyMaxActiveBlocksPerMultiprocessor(
-        &data.max_blocks, func, num_threads, shmem_size));
+        &data.func_max_blocks_per_sm, func, func_threads_per_block, func_dynamic_shmem_per_block));
 #else
     RAJA_UNUSED_VAR(func);
-    data.max_blocks = hip::device_prop().maxThreadsPerMultiProcessor/1024;
-    if (data.max_blocks <= 0) { data.max_blocks = 1 }
+    data.func_max_blocks_per_sm = hip::device_prop().maxThreadsPerMultiProcessor/1024;
+    if (data.func_max_blocks_per_sm <= 0) { data.func_max_blocks_per_sm = 1 }
 #endif
 
-    if (data.multiProcessorCount == uninitialized) {
-
-      data.multiProcessorCount = hip::device_prop().multiProcessorCount;
-
-    }
-
-    data.max_blocks *= data.multiProcessorCount;
-
-    data.prev_shmem_size  = shmem_size;
-    data.prev_num_threads = num_threads;
-
   }
 
-  max_blocks = data.max_blocks;
-
+  return data;
 }
 
-struct HipOccupancyDefaults
+
+/*!
```

Comment:
Is this class identical to the one defined in MemUtils_HIP? Can it be factored out into one implementation that can be used with either CUDA or HIP?

Reply:
That is possible.

Reply:
Does it only apply to forall? What about pattern/detail/reduce.hpp?

Reply:
No this isn't used by reduce, it is currently only used by forall, but it could potentially be used by other things which is why I put it in memutils.

Reply:
Maybe put it in it's own file in the detail directory and include that in the CUDA and HIP files that need it?

Reply:
I'll do that.

Reply:
These call some of the occupancy methods that are cuda/hip specific, I think I'll leave them here for now.

Reply:
Sounds good.